Components/LearningCore/Source/gov/sandia/cognition/learning/algorithm/clustering/OptimizedAffinityPropagation.java

```diff
@@ -0,0 +1,318 @@
+  /*
+   * File:                OptimizedAffinityPropagation.java
+   * Authors:             Marco Pezzulla
+   * Company:             PTV SISTeMA
+   * Project:             Cognitive Foundry
+   *
+   * Copyright August 7, 2007, Sandia Corporation.  Under the terms of Contract
+   * DE-AC04-94AL85000, there is a non-exclusive license for use of this work by
+   * or on behalf of the U.S. Government. Export of this program may require a
+   * license from the United States Government. See CopyrightHistory.txt for
+   * complete details.
+   *
+   */
+
+
+  package gov.sandia.cognition.learning.algorithm.clustering;
+
+  import gov.sandia.cognition.learning.algorithm.clustering.AffinityPropagation;
+  import gov.sandia.cognition.math.DivergenceFunction;
+  import java.util.ArrayList;
+  import java.util.Arrays;
+  import java.util.HashMap;
+  import java.util.stream.DoubleStream;
+
+  /**
+   * The <code>OptimizedAffinityPropagation</code> extends <code>AffinityPropagation</code>.
+   * The <code>OptimizedAffinityPropagation</code> algorithm requires three parameters:
+   * a divergence function, a value to use for self-divergence, and a damping
+   * factor (called lambda in the paper; 0.5 is the default). It clusters by
+   * passing messages between each point to determine the best exemplar for the
+   * point.
+   * <BR><BR>
+   * This implementation takes a divergence function instead of a similarity
+   * function and sets the similarity value to the negative of the divergence
+   * value, as described in the paper for Euclidean distance.
+   * <BR><BR>
+   * The self-divergence value is what controls how many clusters are generated.
+   * Typically this value is set to the mean or median of all the divergence
+   * values or the maximum divergence. In general, a smaller value will mean more
+   * clusters and a larger value will mean less clusters. In the paper this is
+   * called self-similarity (s(k,k)) but since this implementation uses a
+   * divergence metric, we use self-divergence instead.
+   * The implementation of <code>OptimizedAffinityPropagation</code> upgrades the implementation
+   * of <code>AffinityPropagation</code> in terms of computation efficiency.
+   * The two core functions are reimplemented. Computational complexity reduced from n^3 to n^2
+   * for both the functions.
+   * Moreover it adds the computation of the median of similarities if the "selfDivergence"
+   * parameter is not specified in input by the client.
+   * @author  Marco Pezzulla
+   * @since   2.0
+   */
+  public class OptimizedAffinityPropagation<T>
+      extends AffinityPropagation<T>
+  {
+
+    private static final long serialVersionUID = 7170610951827891277L;
+
+    OptimizedAffinityPropagation(){this(null, DEFAULT_SELF_DIVERGENCE);}

```

Comment:
Foundry standards would make this default constructor public since algorithms are generally "Java Beans". Also the formatting could be tweaked.

Reply:
Done